### DISCUSSION TITLE
Event URL with Docker Container Trouble

### DISCUSSION THREAD

**Discussion by bterps** [jason0x43/hacs-hubitat#274] [Q&A] [answered]
Hi! First off, thanks so much for creating and maintaining this add-on! Its people like you that make open source software so great!!

I'm struggling to figure out how to make HA & HE share device updates. I have seen a couple other threads but haven't found a solution that works for me. I have 3 Aqara Temp & Humidty devices that are working fine from a HE perspective, I can see all the ~hourly updates in the logs, but that data isn't making it into HA, it only updates if I change something or restart instances.

The readme calls this out as a known thing to contend with, however I'm really struggling to solve it although i generally understand what its trying to get me to do. Specifics are HE & HA are on the same LAN, but HA is in a docker container so has a different local subnet from HE. I am also running a Caddy proxy and so am forwarding HA traffic. 

HE: 192.168.88.133
HA: 192.168.88.238 / 172.19.0.XX port 8123

When I try to add the 8123 port to the HE integration config in HA, it crashes the instance. I tried adding a 2nd port forward (8124) to see if that would work but no luck.

HE Config:
URL for POST events is currently set to: `http://192.168.88.238`
Platform version: 2.3.8.140

HA Config:
Hub Address: `192.168.88.133`
Event Server URL: `http://192.168.88.238`
Core Version: 2024.5.1
Integration Version: 0.9.28

docker-compose:
```
home-assistant:
container_name: home-assistant
image: "ghcr.io/home-assistant/home-assistant:stable"
environment:
    - PUID=$PUID
    - PGID=$PGID
    - TZ=$TZ
volumes:
    - $DOCKERDIR/home-assistant:/config
    - /etc/localtime:/etc/localtime:ro
    - /run/dbus:/run/dbus:ro
ports:
    - $HOMEASSISTANT_PORT:8123
    - $HA_TO_HUBITAT_PORT:8124
restart: always
privileged: true
```
Caddyfile:
```
home-assistant.home.bradterpstra.com {
reverse_proxy home-assistant:8123
reverse_proxy home-assistant:8124
}
```

**Comment by bterps** [comment]
Let me know what logs or other things are helpful/necessary. Here is a snippet of the log from one of the sensors from HE as well as the graph in HA showing only 1 temp change registered in the last 24 hours. 
```
2024-05-08 01:58:55.974 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor presense check in 12 hours
[dev:16](http://192.168.88.133/logs#)2024-05-08 01:58:55.955 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor presense check in 12 hours
[dev:16](http://192.168.88.133/logs#)2024-05-08 01:58:55.955 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor presense check in 12 hours
[dev:16](http://192.168.88.133/logs#)2024-05-08 01:58:55.851 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor pressure changed to 1007
[dev:16](http://192.168.88.133/logs#)2024-05-08 01:58:55.846 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor humidity changed to 52.94
[dev:16](http://192.168.88.133/logs#)2024-05-08 01:58:55.836 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor temperature changed to 68.414°F
[dev:16](http://192.168.88.133/logs#)2024-05-08 01:39:40.985 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor presense check in 12 hours
[dev:16](http://192.168.88.133/logs#)2024-05-08 01:39:40.966 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor presense check in 12 hours
[dev:16](http://192.168.88.133/logs#)2024-05-08 01:39:40.966 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor presense check in 12 hours
[dev:16](http://192.168.88.133/logs#)2024-05-08 01:39:40.924 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor pressure changed to 1007
[dev:16](http://192.168.88.133/logs#)2024-05-08 01:39:40.909 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor temperature changed to 67.802°F
[dev:16](http://192.168.88.133/logs#)2024-05-08 01:39:40.908 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor humidity changed to 53.37
[dev:16](http://192.168.88.133/logs#)2024-05-08 01:18:25.088 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor presense check in 12 hours
[dev:16](http://192.168.88.133/logs#)2024-05-08 01:18:25.038 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor voltage changed to 2.96 volts
[dev:16](http://192.168.88.133/logs#)2024-05-08 01:18:25.035 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor battery changed to 95%
[dev:16](http://192.168.88.133/logs#)2024-05-08 01:16:04.322 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor presense check in 12 hours
[dev:16](http://192.168.88.133/logs#)2024-05-08 01:16:04.315 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor presense check in 12 hours
[dev:16](http://192.168.88.133/logs#)2024-05-08 01:16:04.310 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor presense check in 12 hours
[dev:16](http://192.168.88.133/logs#)2024-05-08 01:16:04.256 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor pressure changed to 1007
[dev:16](http://192.168.88.133/logs#)2024-05-08 01:16:04.236 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor temperature changed to 66.884°F
[dev:16](http://192.168.88.133/logs#)2024-05-08 01:16:04.236 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor humidity changed to 53.8
[dev:16](http://192.168.88.133/logs#)2024-05-08 01:00:30.142 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor presense check in 12 hours
[dev:16](http://192.168.88.133/logs#)2024-05-08 01:00:30.138 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor presense check in 12 hours
[dev:16](http://192.168.88.133/logs#)2024-05-08 01:00:30.132 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor presense check in 12 hours
[dev:16](http://192.168.88.133/logs#)2024-05-08 01:00:30.108 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor pressure changed to 1007
[dev:16](http://192.168.88.133/logs#)2024-05-08 01:00:30.102 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor temperature changed to 66.29°F
[dev:16](http://192.168.88.133/logs#)2024-05-08 01:00:30.101 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor humidity changed to 54.21
[dev:16](http://192.168.88.133/logs#)2024-05-08 12:41:05.096 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor presense check in 12 hours
[dev:16](http://192.168.88.133/logs#)2024-05-08 12:41:05.085 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor presense check in 12 hours
[dev:16](http://192.168.88.133/logs#)2024-05-08 12:41:05.082 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor presense check in 12 hours
[dev:16](http://192.168.88.133/logs#)2024-05-08 12:41:05.019 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor pressure changed to 1008
[dev:16](http://192.168.88.133/logs#)2024-05-08 12:41:04.978 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor temperature changed to 65.57°F
[dev:16](http://192.168.88.133/logs#)2024-05-08 12:41:04.978 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor humidity changed to 54.87
[dev:16](http://192.168.88.133/logs#)2024-05-08 12:25:20.300 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor presense check in 12 hours
[dev:16](http://192.168.88.133/logs#)2024-05-08 12:25:20.258 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor voltage changed to 2.96 volts
[dev:16](http://192.168.88.133/logs#)2024-05-08 12:25:20.256 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor battery changed to 95%
[dev:16](http://192.168.88.133/logs#)2024-05-08 12:21:29.159 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor presense check in 12 hours
[dev:16](http://192.168.88.133/logs#)2024-05-08 12:21:29.100 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor presense check in 12 hours
[dev:16](http://192.168.88.133/logs#)2024-05-08 12:21:29.081 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor presense check in 12 hours
[dev:16](http://192.168.88.133/logs#)2024-05-08 12:21:29.049 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor pressure changed to 1008
[dev:16](http://192.168.88.133/logs#)2024-05-08 12:21:28.997 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor humidity changed to 55.35
[dev:16](http://192.168.88.133/logs#)2024-05-08 12:21:28.992 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor temperature changed to 64.868°F
[dev:16](http://192.168.88.133/logs#)2024-05-08 12:00:33.403 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor presense check in 12 hours
[dev:16](http://192.168.88.133/logs#)2024-05-08 12:00:33.388 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor presense check in 12 hours
[dev:16](http://192.168.88.133/logs#)2024-05-08 12:00:33.388 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor presense check in 12 hours
[dev:16](http://192.168.88.133/logs#)2024-05-08 12:00:33.372 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor pressure changed to 1008
[dev:16](http://192.168.88.133/logs#)2024-05-08 12:00:33.353 PM[info](http://192.168.88.133/logs#)Master Bedroom Temp Sensor humidity changed to 56.02
```
![image](https://github.com/jason0x43/hacs-hubitat/assets/6097738/0772a588-05ff-4d73-b34a-4104f1ebc973)

**Comment by jason0x43** [comment]
The overall goal is for Hubitat to end up with an address (including a port number) that it can use to post events to the event server started by the integration. The event server is running on the same machine as Home Assistant itself, so the event server will have HA's IP address (or whatever address Hubitat can use to get to HA).

By default, the event server will run at a random port. This is fine when the integration is able to configure things itself -- it will auto-discover its public IP, and tell Hubitat to use `http://<public ip>:<server port>`. Unfortunately, that doesn't work when the integration can't see its public IP, which is the case when HA is running behind a proxy or on a private subnet.

In that case, you need to configure both the event server URL _and_ the event server port.
 
<img width="254" alt="Screenshot 2024-05-07 at 21 11 01" src="https://github.com/jason0x43/hacs-hubitat/assets/1042866/23eba143-332b-4a80-853a-8eb8780a9be4">

The URL is what the integration will tell Hubitat to use to post events. The port value tells the integration to always bind the event server to a specific port rather than choosing a random one. The URL is a separate config option from the port (rather than the URL always using whatever value is given for "Event server port") to allow for port mapping.

**Comment by bterps** [comment]
Thanks @jason0x43 so I think it may have just clicked that I'm setting the event listener port in HA in the HE addon configuration? So I just picked 10484 at random, added it as the event server port in HA, and added it as a forwarded docker port in my docker-compose. Does that sound about right? 
I restarted the HA instance and it has updated in HE to be 10484, so now I can wait a couple hours and see if the temps start updating more frequently. I feel like I was down this path before but i've tried so many different combinations its hard to remember what I tried and what I haven't.
![image](https://github.com/jason0x43/hacs-hubitat/assets/6097738/a3ba8298-1e93-473b-9b42-c74c9dcb0d17)

**Comment by jason0x43** [comment]
Yup. The port you pick in the integration configuration will be used by the event server started by the integration. So when you pick port 10484, the integration is going to start a server on the HA host listening on port 10484.

**Comment by bterps** [comment]
Sadly this didn't work. Got updates due to the changes but haven't had anything since i made the changes yesterday. Are there any logs I can pull to try to figure out why its not working?
![image](https://github.com/jason0x43/hacs-hubitat/assets/6097738/192fbd46-4a50-4efa-9d50-275d3ec063cb)

**Comment by jason0x43** [comment]
Are you sure that the device is trying to report updates? Do any other devices from Hubitat work, particularly ones that generate immediate updates like switches, work?

**Comment by bterps** [comment]
The data is coming into Hubitat, I suppose its possible the Maker API isn't pushing the updates, but the config for that seemed pretty straighforward?
Attached are the Hubitat logs for the Maker API, I'm not sure what it should look like when its trying to push an update, is there a POST event that gets logged? I have "Enable Logging for Debugging" turned on.
[May9-Hubitat-Log.txt](https://github.com/jason0x43/hacs-hubitat/files/15257796/May9-Hubitat-Log.txt)

**Comment by jason0x43** [comment]
Have you tried some other device in Hubitat? Maybe try setting up a virtual contact sensor in Hubitat, add it to the Maker API, and try changing its state in Hubitat and see if it updates in HA.

**Comment by bterps** [comment, marked as answer]
I'm going to close this out as its started working, but sadly I do not know what the actual solution was. I updated Hubitat, the Maker API, this integration, and moved HA from the `Stable` version to `Latest`. No of those were big jumps, I think I was only 1 minor version behind on everything, but guess it was in there somewhere 🤷‍♂️ I did also enable `Allow Access via Remote/Cloud` b/c I was thinking that since they're on different IP subnets maybe was causing issues? Didn't make much sense b/c the integration was able to update the listener port, but getting desperate.

For completeness, this is now the current, working state:
HA
- Core 2024.5.3
- Frontend 20240501.1
- Integration 0.9.28
Hubitat
- Platform 2.3.8.140

**Comment by bterps** [comment]
Very much appreciate your efforts to get me sorted @jason0x43 🙏
